plugins/upower/fu-plugin-upower.c

```diff
@@ -27,128 +25,131 @@ void
+	}
 	if (battery_str == NULL)
-		data->minimum_battery = MINIMUM_BATTERY_PERCENTAGE_FALLBACK;
+		minimum_battery = MINIMUM_BATTERY_PERCENTAGE_FALLBACK;
 	else
-		data->minimum_battery = fu_common_strtoull (battery_str);
-	if (data->minimum_battery > 100) {
-		g_warning ("Invalid minimum battery level specified: %" G_GUINT64_FORMAT,
-			   data->minimum_battery);
-		data->minimum_battery = MINIMUM_BATTERY_PERCENTAGE_FALLBACK;
+		minimum_battery = fu_common_strtoull (battery_str);
+	if (minimum_battery > 100) {
+		g_warning ("invalid minimum battery level specified: %" G_GUINT64_FORMAT,
+			   minimum_battery);
+		minimum_battery = MINIMUM_BATTERY_PERCENTAGE_FALLBACK;
 	}
+	fu_context_set_battery_threshold (ctx, minimum_battery);
+	fu_plugin_upower_rescan (plugin);
 
+	/* success */
 	return TRUE;
 }
 
-static gboolean
-fu_plugin_upower_check_percentage_level (FuPlugin *plugin)
-{
-	FuPluginData *data = fu_plugin_get_data (plugin);
-	gdouble level;
-	guint power_type;
-	g_autoptr(GVariant) percentage_val = NULL;
-	g_autoptr(GVariant) type_val = NULL;
-
-	/* check that we "have" a battery */
-	type_val = g_dbus_proxy_get_cached_property (data->display_proxy, "Type");
-	if (type_val == NULL) {
-		g_warning ("Failed to query power type, assume AC power");
-		return TRUE;
-	}
-	power_type = g_variant_get_uint32 (type_val);
-	if (power_type != 2) {
-		g_debug ("Not running on battery (Type: %u)", power_type);
-		return TRUE;
-	}
-
-	/* check percentage high enough */
-	percentage_val = g_dbus_proxy_get_cached_property (data->display_proxy, "Percentage");
-	if (percentage_val == NULL) {
-		g_warning ("Failed to query power percentage level, assume enough charge");
-		return TRUE;
-	}
-	level = g_variant_get_double (percentage_val);
-	g_debug ("System power source is %.1f%%", level);
-
-	return level >= data->minimum_battery;
-}
-
-static gboolean
-fu_plugin_upower_check_on_battery (FuPlugin *plugin)
-{
-	FuPluginData *data = fu_plugin_get_data (plugin);
-	g_autoptr(GVariant) value = NULL;
-
-	value = g_dbus_proxy_get_cached_property (data->upower_proxy, "OnBattery");
-	if (value == NULL) {
-		g_warning ("failed to get OnBattery value, assume on AC power");
-		return FALSE;
-	}
-	return g_variant_get_boolean (value);
-}
-
 gboolean
 fu_plugin_update_prepare (FuPlugin *plugin,
 			  FwupdInstallFlags flags,
 			  FuDevice *device,
 			  GError **error)
 {
+	FuContext *ctx = fu_plugin_get_context (plugin);
+
 	/* not all devices need this */
 	if (!fu_device_has_flag (device, FWUPD_DEVICE_FLAG_REQUIRE_AC))
 		return TRUE;
+	if (flags & FWUPD_INSTALL_FLAG_IGNORE_POWER)
+		return TRUE;
 
-	/* determine if operating on AC or battery */
-	if (fu_plugin_upower_check_on_battery (plugin) &&
-	    (flags & FWUPD_INSTALL_FLAG_IGNORE_POWER) == 0) {
+	/* not charging */
+	if (fu_context_get_battery_state (ctx) == FU_CONTEXT_BATTERY_STATE_DISCHARGING ||
+	    fu_context_get_battery_state (ctx) == FU_CONTEXT_BATTERY_STATE_EMPTY) {
```

Comment:
can't battery state also be empty if the system is a desktop or doesn't support a battery?

Reply:
If there's no battery then the state would be `FU_CONTEXT_BATTERY_STATE_UNKNOWN`.